extensions/script-libraries/drawing/drawing.livecodescript

```diff
@@ -0,0 +1,2960 @@
+	 * is threaded through all the recursive processing functions, holding
+	 * information about what is being processed (for error handling) and also
+	 * any state which is computed as the processing phrases progress. */
+	local tContext
+	put empty into tContext
+
+	/* First process the svg array, transforming it to a state it can be
+	 * compiled. Each process operation takes tArray as a reference parameter
+	 * so they can update the array in place. */
+
+	/* Remove all elements from the tree which are not understood. */
+	_svgTrim tContext, tArray
+
+	/* Parse the values of all features, including those specified in any inline
+	 * style attributes. Any features which are not applicable to the specifying
+	 * element are removed, and any required attributes are set with their
+	 * default values (required, unset properties are set during cascade). */
+	_svgParse tContext, tArray
+
+	/* Create a map from element id attribute to label. */
+	_svgMap tContext, tArray
+
+	/* Flatten all 'use' elements, replacing them with their referenced
+	 * elements. */
+	_svgFlatten tContext, tArray
+
+	/* Perform the property cascade, assigning appropriate inherited values
+	 * to all properties which apply to a given element. */
+	local tPropertyDefaults
+	put svgSpecGetPropertyDefaults() into tPropertyDefaults
+	repeat for each key tPropertyName in tPropertyDefaults
+		get _svgParseFeatureValue(tContext, "", tPropertyName, tPropertyDefaults[tPropertyName])
+	end repeat
+	_svgCascade tContext, tPropertyDefaults, tArray
+
+	/* Now all processing has been done we can perform the final phase of
+	 * compiling the svg document. */
+
+	/* Compile the processed array, converting its tree structure into a linear
+	 * sequence of graphics operations. */
+	_svgCompile tContext, tArray
+
+
+	/* Encode the operations array in a form suitable for processing by LCB */
+	local tDrawing
+	_svgEncode tContext, tDrawing
+
+	return tDrawing
+end svgImportFromArray
+
+/*******************************************************************************
+ *
+ *  SVG PROCESS OPERATIONS
+ *
+ ******************************************************************************/
+
+/* _svgTrim iterates over the document array, removing any elements which are
+ * are not understood. */
+private command _svgTrim @xContext, @xElement
+	local tElementType
+	put xElement["type"] into tElementType
+
+	/* Push the node's path segment onto the current context so errors can be
+	 * adequately placed. */
+	_svgContextEnter xContext, xElement
+
+	/* If the element type is known, we recurse to its children then process
+	 * the resulting sequence to remove holes */
+	if tElementType is empty then
+		put empty into xElement
+	else if svgSpecIsElement(tElementType) then
+		if xElement["content"] is an array then
+			local tNewContent
+			repeat for each element tChildElement in xElement["content"]
+				/* If a node type is unknown, then _svgProcessNodes will set
```

Comment:
does this mean `_svgTrim`?

Reply:
And similarly `tChildNode` / `tChildElement`?

Reply:
Done